### DISCUSSION TITLE
RSC async Component and writing unit tests

### DISCUSSION THREAD

**Discussion by alecla** [vercel/next.js#46528] [Help] [answered]
I have a RSC component that is `async` since it fetches data. How would i go about to test this in a unit test with `'@testing-library/react`

```js
		render(
				<Suspense fallback={null}>
					{/* @ts-expect-error Server Component */}
					<AsyncComponent {...props} />
				</Suspense>
		);

```

Doesn't work, i get the error:

```js
Error: Objects are not valid as a React child (found: [object Promise]). If you meant to render a collection of children, use an array instead.
````

**Comment by icyJoseph** [comment, marked as answer]
Yeah, this one needs changes upstream I think. The SSR step within Server Components uses react-dom/server methods. While the typical render you find on these libraries is doing a CSR kind of render.

One thing you could do is this suggestion: https://github.com/testing-library/react-testing-library/issues/561#issuecomment-1189796200

```tsx
import { render, screen } from "@testing-library/react";
import { renderToString } from "react-dom/server";

const renderStatic = (
	...[ui, options]: Parameters<typeof render>
): ReturnType<typeof render> => {
	const staticMarkup = renderToString(ui);

	return render(
		<div dangerouslySetInnerHTML={{ __html: staticMarkup }} />,
		options,
	);
};

test("sample static test", () => {
	renderStatic(<button>Click me!</button>);

	expect(screen.queryByRole("button")).toHaveTextContent("Click me!");
});
```

This is probably not enough though, and it is hacky AF... 

The current meta was enough when we could pretend like the SSR, and hydration + CSR were one of the same, under testing environments. We do in-fact render into the JSDOM, IIRC.

A better approach would be to do like Next.js does their tests, but that still would need some more "upstream" work, where whose upstream is not super clear.

Basically they spin up a server and through a browser testing tool, such as Playwright, etc, they see that the rendered content is correct:

- https://github.com/vercel/next.js/blob/canary/test/e2e/app-dir/rsc-basic/rsc-basic.test.ts

That's a far better test for something has so many moving parts though.

- https://github.com/microsoft/playwright/issues/19399

**Reply by alecla** [reply]
Yeah seems like testing through a browser is the way to go currently.
